### PR TITLE
Ubuntu enable unix

### PR DESCRIPTION
**Pull request by alanmcanonical**
#### Description:

- Create bash_pam_unix_enable macro
- Use pam-auth-update to remediate the accounts_password_pam_unix_remember and set_password_hashing_algorithm_systemauth
- In oval, change from any character (.) to non-comment character ([^#]) and add word boundary to pam_unix_algorithms

#### Rationale:

- Use pam-auth-update to dynamically remediate and test pam stack
- The original oval will fail if there is a algo name sha5122, so we need word boundary to constrain the pam_unix_algorithms pattern

### CODE REVIEW

File: linux_os/guide/system/accounts/accounts-restrictions/password_storage/no_empty_passwords/bash/shared.sh [hunk 1]
```diff
@@ -11,8 +11,9 @@ for FILE in ${NULLOK_FILES}; do
    sed --follow-symlinks -i 's/\<nullok\>//g' ${FILE}
 done
 {{% elif 'ubuntu' in product %}}
-FILE="/etc/pam.d/common-password"
-sed -i 's/\(.*pam_unix\.so.*\)\s\<nullok\>\(.*\)/\1\2/g' ${FILE}
+{{{ bash_pam_unix_enable() }}}
+sed --follow-symlinks -i 's/\<nullok\>//g' /usr/share/pam-configs/cac_unix
```
Comment (mpurg) [inline linux_os/guide/system/accounts/accounts-restrictions/password_storage/no_empty_passwords/bash/shared.sh:15]: no_empty_passwords rule is also used in Ubuntu STIGs, which state to remove `nullok` only from `/etc/pam.d/common-password`. https://www.stigviewer.com/stig/canonical_ubuntu_22.04_lts/2024-05-30/finding/V-260570

I suggest that for backwards compatibility, we only remove nullok from the `Passwords` section in `cac_unix`.

File: shared/macros/10-bash.jinja [hunk 2]
```diff
@@ -847,6 +847,45 @@ fi
...
+conf_name=cac_unix
+if [ ! -f /usr/share/pam-configs/"$conf_name" ]; then
+    cat << EOF > /usr/share/pam-configs/"$conf_name"
+Name: Unix authentication
+Default: yes
+Priority: 257
+Conflicts: unix
+Auth-Type: Primary
+Auth:
+        [success=end default=ignore]    pam_unix.so try_first_pass
```
Comment (mpurg) [inline shared/macros/10-bash.jinja:864]: This line should contain nullok, same as in the default unix file.
Rule `no_empty_passwords` or similar should take care of removing the nullok.

File: shared/macros/10-bash.jinja [hunk 3]
```diff
@@ -847,6 +847,45 @@ fi
...
+Auth-Initial:
+        [success=end default=ignore]    pam_unix.so
```
Comment (mpurg) [inline shared/macros/10-bash.jinja:866]: Same as above.

File: shared/macros/10-bash.jinja [hunk 4]
```diff
@@ -847,6 +847,45 @@ fi
...
 
+{{#
+    Enable pam_unix.so PAM module by using pam-auth-update.
+    This option is only recommended when pam-auth-update tool is available for the system.
+#}}
+{{%- macro bash_pam_unix_enable() -%}}
+conf_name=cac_unix
+if [ ! -f /usr/share/pam-configs/"$conf_name" ]; then
+    cat << EOF > /usr/share/pam-configs/"$conf_name"
+Name: Unix authentication
```
Comment (mpurg) [inline shared/macros/10-bash.jinja:858]: Should we create a copy of `unix` (and add Conflicts and bump Priority) instead of hardcoding like this?

**Review by mpurg — CHANGES_REQUESTED** [review]
Changes needed

File: linux_os/guide/system/accounts/accounts-restrictions/password_storage/no_empty_passwords/bash/shared.sh (hunk 1 above)
Comment (alanmcanonical) [reply linux_os/guide/system/accounts/accounts-restrictions/password_storage/no_empty_passwords/bash/shared.sh]: Resolved by https://github.com/ComplianceAsCode/content/blob/aa92d0eb8330d90c8ff3e4ac38d0c3cbfaf03fd5/shared/macros/10-bash.jinja#L864-L866

File: shared/macros/10-bash.jinja (hunk 3 above)
Comment (alanmcanonical) [reply shared/macros/10-bash.jinja]: Resolved by https://github.com/ComplianceAsCode/content/blob/aa92d0eb8330d90c8ff3e4ac38d0c3cbfaf03fd5/shared/macros/10-bash.jinja#L864-L866

File: linux_os/guide/system/accounts/accounts-restrictions/password_storage/no_empty_passwords/bash/shared.sh [hunk 5]
```diff
@@ -11,8 +11,21 @@ for FILE in ${NULLOK_FILES}; do
    sed --follow-symlinks -i 's/\<nullok\>//g' ${FILE}
 done
 {{% elif 'ubuntu' in product %}}
-FILE="/etc/pam.d/common-password"
-sed -i 's/\(.*pam_unix\.so.*\)\s\<nullok\>\(.*\)/\1\2/g' ${FILE}
+{{{ bash_pam_unix_enable() }}}
+config_file="/usr/share/pam-configs/unix"
```
Comment (mpurg) [inline linux_os/guide/system/accounts/accounts-restrictions/password_storage/no_empty_passwords/bash/shared.sh:15]: I think this should be `cac_unix`

File: linux_os/guide/system/accounts/accounts-pam/set_password_hashing_algorithm/set_password_hashing_algorithm_systemauth/oval/shared.xml [hunk 6]
```diff
@@ -1,24 +1,24 @@
...
-    {{% set line_pattern = "^[\s]*password[\s]+(?:\[success=\d+\s+default=ignore\])[\s]+pam_unix\.so[\s]+" %}}
-  {{% else %}}
-    {{% set pam_file = "/etc/pam.d/system-auth" %}}
-    {{% set line_pattern = "^[\s]*password[\s]+(?:(?:required)|(?:sufficient))[\s]+pam_unix\.so[\s]+" %}}
-  {{% endif %}}
-
-  {{% set pam_unix_algorithms = "(sha512|yescrypt|gost_yescrypt|blowfish|sha256|md5|bigcrypt)" %}}
-  {{% set hashing_pattern = line_pattern + "(?!.*" + pam_unix_algorithms + ".*" + pam_unix_algorithms + ").*" + pam_unix_algorithms + ".*$" %}}
+  {{% set pam_unix_algorithms = "\\b(sha512|yescrypt|gost_yescrypt|blowfish|sha256|md5|bigcrypt)\\b" %}}
+  {{% set hashing_pattern = line_pattern + "(?!.*" + pam_unix_algorithms + "[^#]*" + pam_unix_algorithms + ")[^#]*" + pam_unix_algorithms + ".*$" %}}
```
Comment (mpurg) [inline linux_os/guide/system/accounts/accounts-pam/set_password_hashing_algorithm/set_password_hashing_algorithm_systemauth/oval/shared.xml:21]: The changes to the regexes affect all platforms. Would be better to make a separate PR for this.

**Review by mpurg — CHANGES_REQUESTED** [review]
Few changes

File: linux_os/guide/system/accounts/accounts-pam/set_password_hashing_algorithm/set_password_hashing_algorithm_systemauth/oval/shared.xml (hunk 6 above)
Comment (alanmcanonical) [reply linux_os/guide/system/accounts/accounts-pam/set_password_hashing_algorithm/set_password_hashing_algorithm_systemauth/oval/shared.xml:21]: A small change from nay character (.) to non-comnet character ([^#]) and add word boundary to pam_unix_algorithms

File: linux_os/guide/system/accounts/accounts-restrictions/password_storage/no_empty_passwords/bash/shared.sh (hunk 5 above)
Comment (alanmcanonical) [reply linux_os/guide/system/accounts/accounts-restrictions/password_storage/no_empty_passwords/bash/shared.sh]: Solved in [4b4714b](https://github.com/ComplianceAsCode/content/pull/12660/commits/4b4714b4879fa7c799fd4091f96af8cb721c8acd)
